File: test_contracts/vulnerable_contract.py
"""
Vulnerable Algorand Smart Contract (PyTeal/Algopy)
This contract contains INTENTIONAL vulnerabilities for testing purposes
DO NOT USE IN PRODUCTION
"""

from algopy import ARC4Contract, UInt64, String, Txn, Global, op, Bytes
from algopy.arc4 import abimethod

class VulnerableContract(ARC4Contract):
    """A deliberately vulnerable contract for security testing"""
    
    # State variables
    balance: UInt64
    owner: String
    secret_key: String
    admin_backdoor: String
    
    def __init__(self) -> None:
        # VULNERABILITY 1: Hardcoded sensitive data
        self.secret_key = String("sk_live_51234567890abcdef")  # CRITICAL: Hardcoded private key
        self.admin_backdoor = String("ADMIN123BACKDOOR456")    # HIGH: Hardcoded backdoor
        self.balance = UInt64(1000000)
        self.owner = Global.creator_address
    
    @abimethod()
    def withdraw(self, amount: UInt64) -> None:
        """VULNERABILITY 2: Missing access control - CRITICAL"""
        # Anyone can withdraw funds - no authorization check
        self.balance -= amount  # VULNERABILITY 3: Unchecked arithmetic - MEDIUM
        
    @abimethod()
    def emergency_withdraw(self) -> None:
        """VULNERABILITY 4: Logic bomb with hardcoded condition - HIGH"""
        # Backdoor for specific address
        if Txn.sender == "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA":
            self.balance = UInt64(0)  # Drain all funds
            op.exit(1)  # VULNERABILITY 5: Unsafe exit operation - LOW
    
    @abimethod()
    def get_random_winner(self) -> UInt64:
        """VULNERABILITY 6: Weak randomness source - HIGH"""
        # Using predictable timestamp for randomness
        random_value = Global.latest_timestamp % 100
        return random_value
    
    @abimethod()
    def unsafe_transfer(self, to_address: String, amount: UInt64) -> None:
        """VULNERABILITY 7: Missing input validation - MEDIUM"""
        # No validation of parameters
        # No check if amount > balance
        # No validation of to_address format
        self.balance -= amount
        # Transfer logic would go here
    
    @abimethod()
    def process_batch(self, iterations: UInt64) -> None:
        """VULNERABILITY 8: Resource exhaustion / Infinite loop - HIGH"""
        i = UInt64(0)
        # Unbounded loop - can cause DoS
        while i < iterations:  # No upper limit check
            self.balance += 1
            # Missing increment - potential infinite loop
            # i += 1  # This line is commented out!
    
    @abimethod()
    def get_user_balance(self, user: String) -> UInt64:
        """VULNERABILITY 9: Unsafe state access - MEDIUM"""
        # Direct state access without existence check
        # This could fail if user doesn't exist in local state
        return UInt64(0)  # Simplified for demo
    
    @abimethod()
    def admin_function(self, new_owner: String) -> None:
        """VULNERABILITY 10: Timestamp dependency - MEDIUM"""
        # Using timestamp for critical business logic
        if Global.latest_timestamp % 2 == 0:  # Manipulable condition
            self.owner = new_owner
    
    @abimethod()
    def calculate_reward(self, base_amount: UInt64, multiplier: UInt64) -> UInt64:
        """VULNERABILITY 11: Integer overflow potential - MEDIUM"""
        # No bounds checking on multiplication
        result = base_amount * multiplier  # Could overflow
        return result
    
    @abimethod()
    def divide_funds(self, divisor: UInt64) -> UInt64:
        """VULNERABILITY 12: Division by zero - MEDIUM"""
        # No check for zero divisor
        result = self.balance / divisor  # Could cause division by zero
        return result
    
    @abimethod()
    def reentrancy_vulnerable(self, external_contract: UInt64) -> None:
        """VULNERABILITY 13: Reentrancy vulnerability - HIGH"""
        # State change after external call
        # In real implementation, this would make external call first
        # then update state, allowing reentrancy attacks
        self.balance -= 100  # State change should be before external call
        # external_call(external_contract)  # Simulated external call
    
    @abimethod()
    def time_locked_function(self) -> None:
        """VULNERABILITY 14: Timestamp manipulation - MEDIUM"""
        # Critical function depends on exact timestamp
        if Global.latest_timestamp == 1640995200:  # Specific timestamp
            self.balance = UInt64(0)  # Dangerous operation
    
    @abimethod()
    def update_contract(self, new_approval_program: Bytes) -> None:
        """VULNERABILITY 15: Insecure upgrade mechanism - CRITICAL"""
        # No access control for contract updates
        # Anyone can update the contract
        pass  # Update logic would go here
    
    @abimethod()
    def debug_function(self, debug_mode: UInt64) -> String:
        """VULNERABILITY 16: Information disclosure - LOW"""
        if debug_mode == 1:
            # Exposing sensitive internal state
            return self.secret_key  # Leaking sensitive data
        return String("OK")
    
    @abimethod()
    def batch_process_unlimited(self, data: String) -> None:
        """VULNERABILITY 17: No rate limiting - MEDIUM"""
        # No limits on batch processing
        # Could be used for DoS attacks
        for i in range(1000000):  # Hardcoded large number
            # Expensive operation
            pass
    
    @abimethod()
    def legacy_function(self) -> None:
        """VULNERABILITY 18: Dead/unreachable code - LOW"""
        return  # Early return makes code below unreachable
        
        # Dead code below
        self.balance += 1000
        self.owner = String("DEAD_CODE")
    
    @abimethod()
    def weak_validation(self, amount: UInt64, signature: String) -> None:
        """VULNERABILITY 19: Weak cryptographic validation - HIGH"""
        # Weak signature validation
        if len(signature) > 0:  # Insufficient validation
            self.balance += amount
    
    @abimethod()
    def race_condition_vulnerable(self, amount: UInt64) -> None:
        """VULNERABILITY 20: Race condition - MEDIUM"""
        # Check-then-act pattern vulnerable to race conditions
        current_balance = self.balance
        # Time gap here allows race conditions
        if current_balance >= amount:
            self.balance = current_balance - amount

File: test_contracts/vulnerable_contract.ts
/**
 * Vulnerable Algorand TypeScript Contract Integration
 * This file contains INTENTIONAL vulnerabilities for testing purposes
 * DO NOT USE IN PRODUCTION
 */

import algosdk from 'algosdk';
import { Algodv2, Transaction, makeApplicationCallTxnFromObject, makePaymentTxnWithSuggestedParamsFromObject } from 'algosdk';

// VULNERABILITY 1: Hardcoded sensitive credentials - CRITICAL
const HARDCODED_PRIVATE_KEY = "your-private-key-here-do-not-hardcode";  // CRITICAL
const HARDCODED_MNEMONIC = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about";  // CRITICAL
const ADMIN_SECRET = "admin123secret456";  // HIGH

// VULNERABILITY 2: Hardcoded network configuration - MEDIUM
const ALGOD_TOKEN = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
const ALGOD_SERVER = "http://localhost";
const ALGOD_PORT = 4001;

// VULNERABILITY 3: Insecure random number generation - HIGH
function generateInsecureRandom(): number {
    return Math.random();  // Weak randomness source
}

// VULNERABILITY 4: Weak session token generation - HIGH
function generateSessionToken(): string {
    return Math.random().toString(36).substring(7);  // Predictable
}

class VulnerableAlgorandContract {
    private algodClient: Algodv2;
    private appId: number;
    
    constructor(appId: number) {
        // VULNERABILITY 5: No input validation - MEDIUM
        this.appId = appId;  // Should validate appId is positive number
        
        // VULNERABILITY 6: Insecure client configuration - MEDIUM
        this.algodClient = new Algodv2(ALGOD_TOKEN, ALGOD_SERVER, ALGOD_PORT);
    }
    
    // VULNERABILITY 7: Missing access control - CRITICAL
    async adminWithdraw(amount: number): Promise<string> {
        // No authentication or authorization checks
        return await this.withdraw(amount);
    }
    
    // VULNERABILITY 8: Unchecked arithmetic operations - MEDIUM
    async withdraw(amount: number): Promise<string> {
        // No bounds checking
        const balance = await this.getBalance();
        const newBalance = balance - amount;  // Could go negative
        
        // VULNERABILITY 9: Missing error handling - MEDIUM
        const suggestedParams = await this.algodClient.getTransactionParams().do();
        
        // VULNERABILITY 10: Unsigned transaction vulnerability - MEDIUM
        const txn = makeApplicationCallTxnFromObject({
            from: "SENDER_ADDRESS_HERE",  // VULNERABILITY 11: Hardcoded address
            appIndex: this.appId,
            onComplete: algosdk.OnApplicationComplete.NoOpOC,
            appArgs: [new Uint8Array(Buffer.from("withdraw")), new Uint8Array(Buffer.from(amount.toString()))],
            suggestedParams
        });
        
        // VULNERABILITY 12: Transaction not signed before submission
        // Should sign transaction before sending
        
        return "transaction_id_placeholder";
    }
    
    // VULNERABILITY 13: Direct state access without validation - LOW
    async getBalance(): Promise<number> {
        try {
            const accountInfo = await this.algodClient.accountInformation("ACCOUNT_ADDRESS").do();
            return accountInfo.amount;
        } catch (error) {
            // VULNERABILITY 14: Information disclosure in error messages - LOW
            console.log("Error getting balance:", error);  // Logs sensitive info
            return 0;
        }
    }
    
    // VULNERABILITY 15: SQL injection equivalent (unsafe string concatenation) - HIGH
    async searchTransactions(userInput: string): Promise<any> {
        // Unsafe string concatenation for API calls
        const query = `transactions?address=${userInput}`;  // No sanitization
        // In real scenario, this could lead to API injection
        return null;
    }
    
    // VULNERABILITY 16: Race condition vulnerability - MEDIUM
    private isProcessing = false;
    
    async processTransaction(amount: number): Promise<void> {
        // Check-then-act race condition
        if (this.isProcessing) {
            throw new Error("Already processing");
        }
        
        // Time gap here allows race conditions
        await new Promise(resolve => setTimeout(resolve, 100));
        
        this.isProcessing = true;
        // Process transaction
        this.isProcessing = false;
    }
    
    // VULNERABILITY 17: Weak cryptographic validation - HIGH
    validateSignature(message: string, signature: string): boolean {
        // Weak validation - just checks if signature exists
        return signature.length > 0;  // Insufficient validation
    }
    
    // VULNERABILITY 18: Timestamp dependency vulnerability - MEDIUM
    async timeBasedOperation(): Promise<boolean> {
        const currentTime = Date.now();
        
        // Critical logic depends on exact timestamp
        if (currentTime % 1000 === 0) {  // Manipulable condition
            return true;
        }
        
        return false;
    }
    
    // VULNERABILITY 19: Resource exhaustion / DoS vulnerability - HIGH
    async batchProcess(items: any[]): Promise<void> {
        // No limit on batch size
        for (const item of items) {  // Could process millions of items
            await this.expensiveOperation(item);
        }
    }
    
    private async expensiveOperation(item: any): Promise<void> {
        // Simulated expensive operation
        await new Promise(resolve => setTimeout(resolve, 100));
    }
    
    // VULNERABILITY 20: Information disclosure - MEDIUM
    async getDebugInfo(): Promise<any> {
        return {
            privateKey: HARDCODED_PRIVATE_KEY,  // Exposing sensitive data
            adminSecret: ADMIN_SECRET,
            internalState: this.getInternalState()
        };
    }
    
    private getInternalState(): any {
        return {
            appId: this.appId,
            processingStatus: this.isProcessing
        };
    }
    
    // VULNERABILITY 21: Insecure deserialization - HIGH
    async processUserData(serializedData: string): Promise<void> {
        try {
            // Unsafe deserialization
            const userData = JSON.parse(serializedData);  // No validation
            await this.processTransaction(userData.amount);
        } catch (error) {
            // Swallow errors silently
        }
    }
    
    // VULNERABILITY 22: Missing rate limiting - MEDIUM
    async publicEndpoint(data: any): Promise<void> {
        // No rate limiting - can be spammed
        await this.processUserData(JSON.stringify(data));
    }
    
    // VULNERABILITY 23: Weak session management - HIGH
    private sessions: Map<string, any> = new Map();
    
    createSession(userId: string): string {
        const sessionId = generateSessionToken();  // Weak token generation
        this.sessions.set(sessionId, {
            userId,
            createdAt: Date.now(),
            // VULNERABILITY 24: No session expiration
        });
        return sessionId;
    }
    
    // VULNERABILITY 25: Missing input sanitization - MEDIUM
    async logUserAction(action: string, details: string): Promise<void> {
        // No sanitization of log data
        console.log(`User action: ${action}, Details: ${details}`);  // Log injection possible
    }
    
    // VULNERABILITY 26: Insecure random for critical operations - HIGH
    generateTransactionNonce(): string {
        return Math.random().toString();  // Predictable nonce
    }
    
    // VULNERABILITY 27: Missing HTTPS enforcement - MEDIUM
    async makeInsecureRequest(url: string): Promise<any> {
        // No HTTPS enforcement
        const response = await fetch(url);  // Could be HTTP
        return response.json();
    }
    
    // VULNERABILITY 28: Weak password validation - MEDIUM
    validatePassword(password: string): boolean {
        return password.length > 3;  // Weak password requirements
    }
    
    // VULNERABILITY 29: Missing CORS protection - LOW
    setupAPIEndpoint(): void {
        // In a real Express.js setup, this would be:
        // app.use(cors({ origin: '*' }));  // Overly permissive CORS
    }
    
    // VULNERABILITY 30: Insufficient logging - LOW
    async criticalOperation(): Promise<void> {
        // No logging of critical operations
        await this.withdraw(1000000);
    }
}

// VULNERABILITY 31: Global state pollution - LOW
let globalAppInstance: VulnerableAlgorandContract;

// VULNERABILITY 32: Unsafe global functions - MEDIUM
function unsafeGlobalWithdraw(amount: number): Promise<string> {
    // No validation of global state
    return globalAppInstance.withdraw(amount);
}

// VULNERABILITY 33: Memory leak potential - LOW
const eventListeners: Function[] = [];

function addListener(callback: Function): void {
    eventListeners.push(callback);
    // No cleanup mechanism for listeners
}

// VULNERABILITY 34: Prototype pollution potential - HIGH
function mergeObjects(target: any, source: any): any {
    for (const key in source) {
        if (key === '__proto__') {
            // VULNERABILITY: Should prevent prototype pollution
            target[key] = source[key];
        } else {
            target[key] = source[key];
        }
    }
    return target;
}

// VULNERABILITY 35: Unsafe eval usage - CRITICAL
function executeUserCode(code: string): any {
    // NEVER do this in production
    return eval(code);  // Code injection vulnerability
}

// Export vulnerable contract
export default VulnerableAlgorandContract;

// VULNERABILITY 36: Exposed sensitive configuration
export const CONFIG = {
    PRIVATE_KEY: HARDCODED_PRIVATE_KEY,
    ADMIN_SECRET: ADMIN_SECRET,
    DATABASE_URL: "mongodb://admin:password123@localhost:27017/vulnerable_db"
};